set-refs-from-tdx-manifest.py

```diff
@@ -0,0 +1,534 @@
+
+    name: str
+    path: str
+    url: str
+    branch: str | None
+
+
+def _is_gitmodule_header(line: str) -> bool:
+    return line.startswith("[submodule")
+
+
+def _parse_one_gm_payload_line(
+    current_line: str, entry: GitmodulesEntry
+) -> GitmodulesEntry:
+    name, _, value = current_line.strip().split(" ")
+    if name == "path":
+        entry.path = value
+        LOG.debug(f"parsed url {value} from line {current_line}")
+        return entry
+    if name == "url":
+        entry.url = value
+        LOG.debug(f"parsed url {value} from line {current_line}")
+        return entry
+    if name == "branch":
+        entry.branch = value
+        LOG.debug(f"parsed branch {value} from line {current_line}")
+        return entry
+    LOG.warning(
+        f"Not parsing gitmodule line {current_line} in context of {entry}, may need to update script"
+    )
+
+
+def parse_one_gitmodule(
+    current_line: str, rest_lines: Iterable[str]
+) -> tuple[str, Iterable[str], GitmodulesEntry]:
+    """Parse the next entry out of gitmodules."""
+    # first, consume lines until the line is a header (we get the current line as of this function
+    # being called as an argument, because things come out of iterators exactly once so the last call
+    # gave us our first line, and we will give the next call its first line)
+    while not _is_gitmodule_header(current_line):
+        current_line = next(rest_lines)
+    # split into ('[submodule', '"layers/meta-freescale"]')
+    _, pathplus = current_line.split(" ")
+    # drop the quotes and closing brackets
+    name = pathplus[1:-2]
+    entry = GitmodulesEntry(name=name, path="", url="", branch=None)
+    current_line = next(rest_lines)
+    try:
+        # until we get the next header, each line is an attribute
+        while not _is_gitmodule_header(current_line):
+            entry = _parse_one_gm_payload_line(current_line, entry)
+            current_line = next(rest_lines)
+    finally:
+        # return the 'current' (aka first of next block) line. if we ran out of file,
+        # that also ends the module, so it's valid to return everything we currently know.
+        return (current_line, rest_lines, entry)
+
+
+def parse_gitmodules(gitmodules_content: str) -> Iterator[GitmodulesEntry]:
+    """Parse a repo's .gitmodules file into some nice coherent information.
+
+    This is a bit of a pain because .gitmodules looks like this:
+
+    [submodule "layers/meta-freescale"]
+        path = layers/meta-freescale
+        url = https://github.com/Freescale/meta-freescale.git
+
+    So you have to do a little stateful parser thing. This is pretty easy when you're tracking
+    the cursor implicitly by using python iterators, luckily. For parsing details see
+    parse_one_gitmodule; consider the gitmodules_lines iterator to be the cursor.
+
+    """
+    gitmodules_lines = iter(gitmodules_content.strip().split("\n"))
+    current_line = next(gitmodules_lines)
+    while True:
```

Comment:
very cool usage of iterators here!

Reply:
i love iterators and generator expressions. they're so good